teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,339 @@
+
+There are several feature requests in the community which can be met if CEL is supported in Tekton Pipeline:
+- [#3591] calls for supporting ANY match behaviour in `WhenExpressions` to execute the task if any `params` is not empty.
+- [#3149] suggests using CEL to split the string instead of adding an additional task to do the work.
+
+Take a look at current [WhenExpressions]:
+
+```yaml
+when:
+  - input: "$(params.path)"
+    operator: in
+    values: ["README.md"]
+```
+
+Current `WhenExpression` requires users to define three fields: `input`, `operator` and `values`. but CEL expressions can achieve the same functionality with a single string. Additionally, the current operator only supports `in` and `notin`, which can be limiting for complex use cases. For example, users will need to figure out [hacky workarounds][#3591] to support any in their case.
+
+For these reasons, `WhenExpressions` would be a good place to start introducing CEL in Tekton Pipelines. This would allow users to write more concise and powerful conditional expressions, and it could be extended to other fields in the future.
+
+### Use Cases
+
+- Cover all the functionalities of current `WhenExpression` and offer a simpler way of having the conditional execution
+- Support ANY: e.g. Execute the task with if any params is not empty
+- Support numeric operators such as GreaterThan/LessThan, e.g. Execute the task if test coverage is large than a certain number.
+
+### Goals
+- Support CEL in `WhenExpression`.
+
+### Non-Goals
+- Support CEL in Pipeline API fields other than `WhenExpression`.
+
+### Related Work
+
+Similar expression evaluations are supported in other CI/CD tools in the industry:
+- [Argo][Argo]: `Workflows` support conditional execution using a when property to specify whether to run or skip a `Step` ([example][argo example]).
+- [GitHub Actions][GitHub Actions]: supports evaluating expressions in `if` of a `step` to enable conditional execution.
+- [Jenkins][Jenkins]: executes the stage when the specified Groovy expression evaluates to true.
+- [Spinnaker][Spinnaker]: Uses string expressions, `Stages` only run when expressions evaluate to `True`.
+
+
+## Proposal
+
+This doc proposes to add a `string` type field `cel` under `WhenExpression`. Like [CelCustomRun][CELCustomRun], we propose to use [cel-go][cel-go] for the expression evaluation. This will allow users to craft any valid CEL expression as defined by the [cel-spec][cel-spec] language definition.
+
+
+### Syntax
+
+In this section we will show how CEL can support the [use cases](#use-cases).
+
+#### Use CEL to replace current WhenExpression
+```yaml
+# current WhenWxpressions
+when:
+  - input: "foo"
+    operator: "in"
+    values: ["foo", “bar”]
+  - input: "duh"
+    operator: "notin"
+    values: ["foo", “bar”]
+
+# with cel
+when:
+  - cel: "\”foo\” in [\"foo\", \“bar\”]”
+  - cel: "\”duh\” not in [\"foo\", \“bar\”]”
```

Comment:
Using a mix of `'` and `"` would work as well, wouldn't it ?

Reply:
Yes, I will fix those